### PR TITLE
Save NotificationToken from Combine call

### PR DESCRIPTION
**Pull request by leemaguire**
Adds the ability to capture a NotificationToken when using a Combine publisher that observes a Realm Object or Collection. The user will call `saveToken(on:at:)` directly after invoking the publisher to use the feature.

### CODE REVIEW

File: RealmSwift/Combine.swift [hunk 1]
```diff
@@ -324,7 +324,7 @@ extension Realm {
     /// A publisher that emits Void each time the object changes.
     ///
     /// Despite the name, this actually emits *after* the collection has changed.
-    public var objectWillChange: Publishers.RealmWillChange {
+    public var objectWillChange: Publishers.RealmWillChange { // Maybe use `Never`? Would using Never confuse users?
```
Comment (jsflax) [inline RealmSwift/Combine.swift:327]: Stray comment

File: RealmSwift/Combine.swift [hunk 2]
```diff
@@ -437,14 +439,14 @@ extension AnyRealmCollection: RealmSubscribable {
 /// A subscription which wraps a Realm notification.
 @available(OSX 10.15, watchOS 6.0, iOS 13.0, iOSApplicationExtension 13.0, OSXApplicationExtension 10.15, tvOS 13.0, *)
 public struct ObservationSubscription: Subscription {
-    private var token: NotificationToken
-    internal init(token: NotificationToken) {
-        self.token = token
+    public var cancellable: NotificationToken
```
Comment (jsflax) [inline RealmSwift/Combine.swift:442]: This requires a docstring now.

File: RealmSwift/Combine.swift [hunk 3]
```diff
@@ -685,37 +852,46 @@ public enum Publishers {
     /// should always be the first operation in the pipeline.
     ///
     /// Create this publisher using the `objectChangeset()` function.
-    public struct ObjectChangeset<T: Object>: Publisher {
+    public struct ObjectChangeset<O: Object>: Publisher {
```
Comment (jsflax) [inline RealmSwift/Combine.swift:855]: Probably could've kept this `T`.

File: RealmSwift/Combine.swift (hunk 2 above)
Comment (leemaguire) [reply RealmSwift/Combine.swift]: changed back to private, no need

File: RealmSwift/Combine.swift (hunk 3 above)
Comment (leemaguire) [reply RealmSwift/Combine.swift:853]: It can but below it is `ObjectChangesetWithToken<O: Object, T>`. Maybe for consistency it should be `O`?

Comment (jsflax) [reply RealmSwift/Combine.swift:853]: Maybe– I'm not sure there is mandated naming convention here. Generally speaking, `T` for type, `U` for the next type (and so on).  `K`/`Key` for key and `V`/`Value`. But if `O` gives you consistency, then I wouldn't be terribly pedantic. As you were.

File: RealmSwift/Combine.swift [hunk 4]
```diff
@@ -503,14 +544,57 @@ public enum Publishers {
         /// This publisher emits Void.
         public typealias Output = Void
 
+        private let collection: Collection
+
+        internal init(_ collection: Collection) {
+            self.collection = collection
+        }
+
+        /// Allows you to capture the `NotificationToken` produced by observing a Realm Collection.
```
Comment (tgoyne) [inline RealmSwift/Combine.swift:553]: This needs a bit more description of what it's doing and why you'd want to use it. It should mention notification skipping, how it makes that possible, and when the token is actually written to the target object/keypath.

File: RealmSwift/Combine.swift [hunk 5]
```diff
@@ -503,14 +544,57 @@ public enum Publishers {
...
+        /// - Parameters:
+        ///   - object: The object which owns the `NotificationToken`.
+        ///   - keyPath: The KeyPath of the `NotificationToken`.
```
Comment (tgoyne) [inline RealmSwift/Combine.swift:556]: The object which the notification token is written to, and the keypath which is it written to.

File: RealmSwift/Combine.swift [hunk 6]
```diff
@@ -503,14 +544,57 @@ public enum Publishers {
...
+        /// - Returns: A `WillChangeWithToken` Publisher.
+        public func saveToken<T>(on object: T, for keyPath: WritableKeyPath<T, NotificationToken?>) -> WillChangeWithToken<Collection, T> {
```
Comment (tgoyne) [inline RealmSwift/Combine.swift:558]: `for:` doesn't make sense here. Maybe `at:`?

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,25 @@
+x.y.z Release notes (yyyy-MM-dd)
+=============================================================
+### Enhancements
+* Add ability to save a NotificationToken when using Combine for observing an Object or Collection by using `saveToken(on:for:)` before `sink` in a chain.
```
Comment (tgoyne) [inline CHANGELOG.md:4]: Using this doesn't require that `.sink` appears in the chain and the position is quite a bit more constrained than that.

File: RealmSwift/Combine.swift [hunk 8]
```diff
@@ -486,10 +486,12 @@ public enum Publishers {
             self.realm = realm
         }
 
-        /// Allows you to capture the `NotificationToken` produced by observing a Realm.
+        /// Captures the `NotificationToken` produced by observing a Realm Collection. This allows you to do notification skipping
```
Comment (tgoyne) [inline RealmSwift/Combine.swift:489]: Paragraph break after the first sentence, as when the docstrings are rendered the first line is used as the short description and then everything after that is formatted a bit differently.

**Comment by tgoyne** [comment]
Needs changelog entries for the bug fix pulled in via the object store upgrade.

**Review by tgoyne — APPROVED** [review]

**Pull request merged**